On #42: Fax pdf/csv: Decimals dependant on supplier_order_unit's si convertability

lentschi opened the pull request:
Make display of decimals dependant on supplier_order_unit's si convertability

lentschi committed:
On #42: Fax pdf/csv: Decimals dependant on supplier_order_unit's si convertability

twothreenine committed:
Solve #42: Improve fax PDF, CSV, text

- outsource format_units_to_order to OrderHelper
- fax text: include unit, adjust column width
- fax PDF & text: only include order number if any present

twothreenine commented on app/lib/order_txt.rb:
```diff
@@ -17,11 +19,39 @@ def to_txt
+    longest_number_string_length = 0
+    longest_amount_string_length = I18n.t('orders.fax.amount').length
+    longest_unit_string_length = I18n.t('orders.fax.unit').length
+    @order_positions = @order.order_articles.ordered.includes(:article_version).map do |oa|
```
Here I'm unsure whether this needs to be an instance variable or if there are any adverse effects of this. I was a bit confused with the for loop in line 52, that's why I used an `@`

lentschi replied:
No, there's no instance variable required here.

lentschi commented on app/lib/order_txt.rb:
```diff
@@ -17,11 +19,39 @@ def to_txt
-                   I18n.t('orders.fax.name'))
+
+    # prepare order_articles data
+    longest_number_string_length = 0
+    longest_amount_string_length = I18n.t('orders.fax.amount').length
+    longest_unit_string_length = I18n.t('orders.fax.unit').length
+    @order_positions = @order.order_articles.ordered.includes(:article_version).map do |oa|
+      number = oa.article_version.order_number || ''
+      amount = format_units_to_order(oa).to_s
+      unit = format_supplier_order_unit_with_ratios(oa.price)
+      longest_number_string_length = number.length if number.length > longest_number_string_length
+      longest_amount_string_length = amount.length if amount.length > longest_amount_string_length
+      longest_unit_string_length = unit.length if unit.length > longest_unit_string_length
+      {
+        number: number,
+        amount: amount,
+        unit: unit,
+        name: oa.article_version.name
+      }
+    end
+
+    if (any_number_present = longest_number_string_length > 0) && longest_number_string_length < I18n.t('orders.fax.number').length
+      longest_number_string_length = I18n.t('orders.fax.number').length
+    end
```
This would do the same with a little less redundancy:

```suggestion
    order_articles =  @order.order_articles.ordered.includes(:article_version)
    order_positions = order_articles.map do |oa|
      number = oa.article_version.order_number || ''
      amount = format_units_to_order(oa).to_s
      unit = format_supplier_order_unit_with_ratios(oa.price)
      {
        number: number,
        amount: amount,
        unit: unit,
        name: oa.article_version.name
      }
    end

    any_number_present = order_articles.any? { |oa| oa.article_version.order_number.present? }
    longest_number_string_length = (order_positions.pluck(:number) + [I18n.t('orders.fax.number')]).map(&:length).max if any_number_present
    longest_amount_string_length = (order_positions.pluck(:amount) + [I18n.t('orders.fax.amount')]).map(&:length).max
    longest_unit_string_length = (order_positions.pluck(:unit) + [I18n.t('orders.fax.unit')]).map(&:length).max
```

It would be even better to make this more generic though, so we wouldn't have to alter the code so much whenever we remove/add a column and to reduce the method's size (too big after the changes!).
This could be done by using something like https://github.com/aptinio/text-table (<- quite antique, so either we find something newer or do our own implementation.)

All this is quite out of scope as far as the hackathon fork is concerned though, and I wonder if we should implement dynamic column widths here at all. (Should probably be discussed upstream instead.)

lentschi commented on app/documents/order_fax.rb:
```diff
@@ -74,33 +75,41 @@ def recipient_paragraph
-      data << [oa.article_version.order_number,
-               quantity,
+      order_number = oa.article_version.order_number
+      any_order_number_present = true if order_number
```
```suggestion
      any_order_number_present = true if order_number.present?
```

lentschi commented on app/documents/order_fax.rb:
```diff
@@ -74,33 +75,41 @@ def recipient_paragraph
   def articles_paragraph
     total = 0
-    data = [I18n.t('documents.order_fax.rows')]
+    data = [I18n.t('documents.order_fax.rows').clone]
```
What's the `clone` for?

lentschi commented on app/documents/order_fax.rb:
```diff
@@ -74,33 +75,41 @@ def recipient_paragraph
+    unless any_order_number_present
+      data.map { |row| row.delete_at(0) }
+      total_row.delete_at(1)
+    end
```
Removing the `order_number` column when no article has an order_number this way is quite quirky.

Also it's out of scope of the hackathon fork: The same could be done upstream at a later point. This just increases complexity for the final PR and all required rebases on master leading up to it. (The same goes for the corresponding logic in `order_txt`)

lentschi commented on app/helpers/orders_helper.rb:
```diff
@@ -23,6 +23,15 @@ def options_for_suppliers_to_select
     options_for_select(options)
   end
 
+  def format_amount(amount, order_article, strip_insignificant_zeros: false)
+    strip_insignificant_zeros = true unless order_article.article_version.supplier_order_unit_is_si_convertible
+    number_with_precision(amount, precision: 3, strip_insignificant_zeros: strip_insignificant_zeros)
+  end
```
- This method is solely used inside the helper -> should be private
- Its name isn't very conclusive - format _how_? (Maybe `format_amount_for_export`?)

lentschi replied:
I changed it to be private, so the name is not so important.

lentschi commented on app/helpers/orders_helper.rb:
```diff
@@ -23,6 +23,15 @@ def options_for_suppliers_to_select
     options_for_select(options)
   end
 
+  def format_amount(amount, order_article, strip_insignificant_zeros: false)
+    strip_insignificant_zeros = true unless order_article.article_version.supplier_order_unit_is_si_convertible
+    number_with_precision(amount, precision: 3, strip_insignificant_zeros: strip_insignificant_zeros)
+  end
+
+  def format_units_to_order(order_article, strip_insignificant_zeros: false)
+    format_amount(order_article.units_to_order, order_article, strip_insignificant_zeros: strip_insignificant_zeros)
+  end
```
Both those new methods might be used in other cases later on and don't really require an `order_article` to work - they should rather take `article_version` as parameter. Subsequently they should be moved to the `ArticlesHelper`, which currently is the designated helper for both `Article` and `ArticleVersion`.

lentschi committed:
On #42:

- Adapted order_txt to generalized creating the text table and added
  spec
- Code style fixes for order_fax

twothreenine commented:
I found an error and posted a solution: https://github.com/foodcoopsat/foodsoft_hackathon/commit/dad0bb90cbf937ff09924d3d237c3d7b9e3db72d#r143648091 (don't know why my comment doesn't show up in this stream)

Otherwise good to go 👍

lentschi committed:
On #42 Fixes error https://github.com/foodcoopsat/foodsoft_hackathon/commit/dad0bb90cbf937ff09924d3d237c3d7b9e3db72d#r143648091
